### PR TITLE
Ensure Proper Order of CLI Build Tasks

### PR DESCRIPTION
**Pull request by dubeejw**
@bpoole16 is experiencing CLI build issues. Attempting to resolve those issues.

### CODE REVIEW

**Comment by dubeejw** [comment]
bpoole16, please test this out and see if the CLI builds and gets copied properly.

**Comment by dubeejw** [comment]
@mdeuser, please review.

**Comment by dubeejw** [comment]
PG3 861, approved.

File: tools/cli/build.gradle [hunk 1]
```diff
@@ -33,17 +30,26 @@ task dumpOSInfo {
...
+            exec {
+                executable "cp"
+                args "${from_path_wsk}", "${to_path_dir}"
+            }
+        }
+    }
 }
 
-pushImage.finalizedBy copyCLIShortcut
+distBinary.dependsOn removeBinary
```
Comment (mdeuser) [inline tools/cli/build.gradle]: previously, distBinary depended on both removeBinary and distDocker.  with this change, the only dependency is on removeBinary.  is distDocker in the dependency chain somewhere?

Comment (dubeejw) [inline tools/cli/build.gradle]: @mdeuser, the dependancies still exist. `distBinary` now runs after `pushImage` which depends on `distDocker`. `distBinary.dependsOn removeBinary` and `pushImage.finalizedBy distBinary`.

**Comment by dubeejw** [comment]
@markusthoemmes, can you review this one?

**Comment by dubeejw** [comment]
@markusthoemmes, any comments?

File: tools/cli/build.gradle [hunk 2]
```diff
@@ -33,17 +30,26 @@ task dumpOSInfo {
...
+        String from_path_wsk = "${projectDir}/../../bin/${go_osname}/${go_osarch}/wsk"
+        String to_path_dir = "${projectDir}/../../bin"
+
+        if (file(from_path_wsk).exists()) {
+            exec {
+                executable "cp"
+                args "${from_path_wsk}", "${to_path_dir}"
+            }
+        }
+    }
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: Why was this change needed?

File: tools/cli/build.gradle [hunk 3]
```diff
@@ -33,17 +30,26 @@ task dumpOSInfo {
     }
 }
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: Can we get rid of this one? Is it needed?

File: tools/cli/build.gradle (hunk 2 above)
Comment (dubeejw) [inline tools/cli/build.gradle]: If I recall correctly the exec functionallity worked better on the Windows machine running Vagrant.

**Comment by dubeejw** [comment]
@markusthoemmes, any more comments on this one?

File: tools/cli/build.gradle [hunk 4]
```diff
@@ -11,39 +11,33 @@ task removeBinary(type: Delete) {
...
-        println "go.arch = "+mapOsArchToGoArch(getOsArch())
+        String go_osname = mapOsNameToGoName(getOsName())
+        String go_osarch = mapOsArchToGoArch(getOsArch())
+        String from_path_wsk = "${projectDir}/../../bin/${go_osname}/${go_osarch}/wsk"
+        String to_path_dir = "${projectDir}/../../bin"
+
+        exec {
+            executable "cp"
+            args "${from_path_wsk}", "${to_path_dir}"
+        }
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: I feel like folding this into `distBinary` could be valuable here vs. having the `finalizedBy` statement below, wdyt? Both are `doLast` blocks so it should just be copy&paste.

File: tools/cli/build.gradle [hunk 5]
```diff
@@ -11,39 +11,33 @@ task removeBinary(type: Delete) {
     delete "${projectDir}/../../bin/windows"
 }
 
-task distBinary(dependsOn: [removeBinary, distDocker]) {
+task distBinary() {
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: Why not keep at least `removeBinary` in this block?

File: tools/cli/build.gradle [hunk 6]
```diff
@@ -11,39 +11,33 @@ task removeBinary(type: Delete) {
...
-    String to_path_dir   = "${projectDir}/../../bin"
-
-    from from_path_wsk
-    into to_path_dir
-}
-
-pushImage.finalizedBy copyCLIShortcut
+distBinary.dependsOn removeBinary
+distBinary.finalizedBy copyCLIShortcut
+pushImage.finalizedBy distBinary
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: If you make the changes suggested above, the `pushImage.finalizedBy distBinary` is the only connection left. I suggest moving that just below the definition of `distBinary` then to keep the context.

**Comment by markusthoemmes** [comment]
Few nits, otherwise LGTM.

**Comment by rabbah** [comment]
@dubeejw can you address the last suggestions?

**Comment by dubeejw** [comment]
@rabbah, comments addressed.

**Comment by dubeejw** [comment]
PG1 2196 ⏳

File: tools/cli/build.gradle [hunk 7]
```diff
@@ -11,22 +11,18 @@ task removeBinary(type: Delete) {
...
 }
 
-task distBinary() {
+task distBinary(dependsOn: [removeBinary]) {
     doLast {
+        String go_osname = mapOsNameToGoName(getOsName())
+        String go_osarch = mapOsArchToGoArch(getOsArch())
+        String from_path_wsk = "${projectDir}/../../bin/${go_osname}/${go_osarch}/wsk"
+        String to_path_dir = "${projectDir}/../../bin"
+
```
Comment (markusthoemmes) [inline tools/cli/build.gradle]: Missing the `exec` to actually copy?

Comment (dubeejw) [reply tools/cli/build.gradle]: The git diff looks strange. exec is in distBinary though.

```
task distBinary(dependsOn: [removeBinary]) {
    doLast {
        String go_osname = mapOsNameToGoName(getOsName())
        String go_osarch = mapOsArchToGoArch(getOsArch())
        String from_path_wsk = "${projectDir}/../../bin/${go_osname}/${go_osarch}/wsk"
        String to_path_dir = "${projectDir}/../../bin"

        // Copy all Go binaries from Docker into openwhisk/bin folder
        run(dockerBinary + ["rm", "-f", dockerContainerName], true)
        run(dockerBinary + ["run", "--name", dockerContainerName, dockerTaggedImageName])
        run(dockerBinary + ["cp", dockerContainerName + ":/src/github.com/go-whisk-cli/build/.", "${projectDir}/../../bin"])
        run(dockerBinary + ["rm", "-f", dockerContainerName])

        exec {
            executable "cp"
            args "${from_path_wsk}", "${to_path_dir}"
        }
    }
}
```
